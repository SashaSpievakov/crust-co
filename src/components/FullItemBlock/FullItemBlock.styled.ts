import styled from 'styled-components';
import { Container } from '../../styles/Base.styled';

const Item = styled(Container)`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;

  @media screen and (max-width: 600px) {
    gap: 15px;
  }

  @media screen and (max-width: 400px) {
    gap: 10px;
  }
`;

const Image = styled.img`
  margin-top: 60px;
  margin-bottom: 40px;
  border-radius: 20px;

  @media screen and (max-width: 600px) {
    width: 300px;
    height: 300px;
    margin-top: 50px;
    margin-bottom: 10px;
  }

  @media screen and (max-width: 400px) {
    width: 200px;
    height: 200px;
    margin-top: 20px;
    margin-bottom: 0;
  }
`;

const Title = styled.h2`
  font-size: 45px;
  font-weight: 700;
  text-align: center;

  @media screen and (max-width: 600px) {
    font-size: 35px;
    margin-bottom: 0;
  }

  @media screen and (max-width: 400px) {
    line-height: 35px;
  }
`;

const Paragraph = styled.p`
  font-size: 19px;
  max-width: 600px;
  text-align: center;
  margin-bottom: 15px;

  @media screen and (max-width: 550px) {
    font-size: 14px;
  }
`;

const Price = styled.h3`
  font-size: 45px;

  @media screen and (max-width: 500px) {
    font-size: 40px;
  }
`;

export { Item, Image, Title, Paragraph, Price };
