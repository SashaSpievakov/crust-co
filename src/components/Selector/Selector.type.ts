export interface ISelector {
  price: number;
  sizes: number[];
  types: number[];
  activeSize: number;
  activeType: number;
  setActivePrice: (index: number) => void;
  setActiveSize: (index: number) => void;
  setActiveType: (index: number) => void;
  typeNames: string[];
  isFullScreen?: boolean;
}
