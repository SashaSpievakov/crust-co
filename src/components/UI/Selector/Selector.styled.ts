import styled, { css } from 'styled-components';

interface LiProps {
  chosen: boolean;
}

interface DivProps {
  isFullScreen: boolean;
}

const Div = styled.div<DivProps>`
  display: flex;
  font-size: ${({ isFullScreen }) => (isFullScreen ? '18px' : '14px')};
  flex-direction: column;
  gap: 6px;
  padding: 6px;
  background-color: ${({ theme }) => theme.colors.gray};
  border-radius: 10px;

  ul {
    display: flex;
    flex: 1;
    text-align: center;
  }

  @media screen and (max-width: 400px) {
    font-size: ${({ isFullScreen }) => (isFullScreen ? '16px' : '14px')};
  }

  ${({ isFullScreen }) =>
    isFullScreen &&
    css`
      min-width: 400px;

      @media screen and (max-width: 550px) {
        min-width: 300px;
      }

      @media screen and (max-width: 400px) {
        min-width: 230px;
      }
    `}
`;

const Li = styled.li<LiProps>`
  font-weight: 600;
  flex: 1;
  padding: 8px;
  cursor: pointer;

  ${({ chosen }) =>
    chosen &&
    css`
      color: ${({ theme }) => theme.colors.black};
      background: ${({ theme }) => theme.colors.chosen};
      box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.04);
      border-radius: 5px;
      cursor: auto;
    `}
`;

export { Div, Li };
