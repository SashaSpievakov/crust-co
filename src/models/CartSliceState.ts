import { ICartItem } from './ICartItem';

export interface CartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
