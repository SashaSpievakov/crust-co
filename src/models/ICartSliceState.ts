import { ICartItem } from './ICartItem';

export interface ICartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
