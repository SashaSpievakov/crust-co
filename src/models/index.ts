export type { ICartItem } from './ICartItem';
export type { IPizzaItem } from './IPizzaItem';
