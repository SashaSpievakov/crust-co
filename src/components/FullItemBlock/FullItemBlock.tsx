import { useState } from 'react';

import { IPizzaItem } from '../../models';
import ItemsCountHandler from '../ItemsCountHandler/ItemsCountHandler';
import Selector from '../Selector/Selector';
import { ButtonLink } from '../UI';
import { Image, Item, Paragraph, Price, Title } from './FullItemBlock.styled';

interface FullItemProps {
  item: IPizzaItem;
}

export const typeNames: string[] = ['traditional', 'thin'];

const FullItemBlock = ({ item }: FullItemProps) => {
  const [activePrice, setActivePrice] = useState<number>(item.price);
  const [activeSize, setActiveSize] = useState<number>(12);
  const [activeType, setActiveType] = useState<number>(0);

  return (
    <Item data-testid="itemPage">
      <picture>
        <source
          srcSet={`../assets/img/pizza${item.id}.webp`}
          type="image/webp"
        />
        <source srcSet={`../assets/img/pizza${item.id}.png`} type="image/png" />
        <Image
          src={`../assets/img/pizza${item.id}.png`}
          alt={`${item.name} pizza`}
          width="450"
          height="450"
        />
      </picture>
      <Title>{item.name}</Title>
      <Paragraph>{item.description}</Paragraph>
      <Selector
        price={activePrice}
        sizes={item.sizes}
        types={item.types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
        isFullScreen
      />
      <Price>{activePrice}$</Price>
      <ItemsCountHandler
        id={item.id}
        name={item.name}
        price={activePrice}
        activeSize={activeSize}
        activeType={activeType}
        typeNames={typeNames}
        isFullScreen
      />
      <ButtonLink link="/">Go Back</ButtonLink>
    </Item>
  );
};

export default FullItemBlock;
