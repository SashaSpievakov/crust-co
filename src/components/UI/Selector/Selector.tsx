import { KeyboardEvent, useLayoutEffect } from 'react';

import { Section, Li } from './Selector.styled';
import { ISelector } from '../../../models/ISelector';

const Selector = ({
  price,
  sizes,
  types,
  activeSize,
  activeType,
  setActivePrice,
  setActiveSize,
  setActiveType,
  typeNames,
  isFullScreen = false,
}: ISelector) => {
  const typeClickHandler = (type: number) => {
    if (activeType === 0 && type === 1) {
      setActivePrice(price + 1);
    } else if (activeType === 1 && type === 0) {
      setActivePrice(price - 1);
    }

    setActiveType(type);
  };

  const sizeClickHandler = (size: number) => {
    if (activeSize === 12 && size === 14) {
      setActivePrice(price + 2);
    } else if (activeSize === 12 && size === 16) {
      setActivePrice(price + 4);
    } else if (activeSize === 14 && size === 16) {
      setActivePrice(price + 2);
    } else if (activeSize === 14 && size === 12) {
      setActivePrice(price - 2);
    } else if (activeSize === 16 && size === 14) {
      setActivePrice(price - 2);
    } else if (activeSize === 16 && size === 12) {
      setActivePrice(price - 4);
    }

    setActiveSize(size);
  };

  const handleKeyDownType = (e: KeyboardEvent<HTMLElement>, type: number) => {
    if (e.code === 'Enter') typeClickHandler(type);
  };

  const handleKeyDownSize = (e: KeyboardEvent<HTMLElement>, size: number) => {
    if (e.code === 'Enter') sizeClickHandler(size);
  };

  useLayoutEffect(() => {
    setActiveSize(sizes[0]);
    setActiveType(types[0]);
  }, [sizes]);

  return (
    <Section isFullScreen={isFullScreen}>
      <ul>
        {types.map((type: number) => (
          <Li
            key={type}
            onClick={() => typeClickHandler(type)}
            onKeyDown={(e) => handleKeyDownType(e, type)}
            chosen={activeType === types.indexOf(type)}
            tabIndex={0}
            aria-current={activeType === types.indexOf(type)}
          >
            {typeNames[type]}
          </Li>
        ))}
      </ul>
      <ul>
        {sizes.map((size: number) => (
          <Li
            key={size}
            onClick={() => sizeClickHandler(size)}
            onKeyDown={(e) => handleKeyDownSize(e, size)}
            chosen={activeSize === size}
            tabIndex={0}
            aria-current={activeSize === size}
          >
            {size} inch
          </Li>
        ))}
      </ul>
    </Section>
  );
};
export default Selector;
