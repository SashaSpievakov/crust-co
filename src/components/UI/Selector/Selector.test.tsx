import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Selector from './Selector';
import rendererWithAllProviders from '../../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../../tests/helpers/renderWithProviders';
import { typeNames } from '../../FullItemBlock/FullItemBlock';
import { ISelector } from '../../../models/ISelector';

const setPriceMock = jest.fn();
const setSizeMock = jest.fn();
const setTypeMock = jest.fn();

const SelectorMockProps: ISelector = {
  price: 10,
  sizes: [12, 14, 16],
  types: [0, 1],
  activeSize: 0,
  activeType: 0,
  setActivePrice: setPriceMock,
  setActiveSize: setSizeMock,
  setActiveType: setTypeMock,
  typeNames,
  isFullScreen: false,
};

describe('Selector Tests', () => {
  test('renders the Selector component', () => {
    const snapshot = rendererWithAllProviders(
      <Selector {...SelectorMockProps} />,
    );
    expect(snapshot).toMatchSnapshot();
  });
});
