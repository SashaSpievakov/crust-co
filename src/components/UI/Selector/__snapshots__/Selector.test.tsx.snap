// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector Tests renders the Selector UI component 1`] = `
Object {
  "children": Array [
    <ul>
      <li
        aria-current={true}
        className="sc-gswNZR iMRIUn"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        traditional
      </li>
      <li
        aria-current={false}
        className="sc-gswNZR kCsqcO"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        thin
      </li>
    </ul>,
    <ul>
      <li
        aria-current={false}
        className="sc-gswNZR kCsqcO"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        12
         inch
      </li>
      <li
        aria-current={false}
        className="sc-gswNZR kCsqcO"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        14
         inch
      </li>
      <li
        aria-current={false}
        className="sc-gswNZR kCsqcO"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        16
         inch
      </li>
    </ul>,
  ],
  "props": Object {
    "className": "sc-bcXHqe fWeZKg",
  },
  "type": "section",
}
`;
